File: golang/intervals/insert_interval.go
package main

func insert(intervals [][]int, new_interval []int) [][]int {
	result := [][]int{}
	i := 0
	n := len(intervals)

	for i < n && intervals[i][1] < new_interval[0] {
		result = append(result, intervals[i])
		i++
	}

	for i < n && intervals[i][0] <= new_interval[1] {
		new_interval[0] = min(new_interval[0], intervals[i][0])
		new_interval[1] = max(new_interval[1], intervals[i][1])

		i++
	}

	result = append(result, new_interval)

	for i < n {
		result = append(result, intervals[i])

		i++
	}

	return result
}

File: python/intervals/insert_interval.py
from typing import List


class Solution:
    def insert(
        self, intervals: List[List[int]], new_interval: List[int]
    ) -> List[List[int]]:
        merged = []
        i = 0
        n = len(intervals)

        while i < n and intervals[i][1] < new_interval[0]:
            merged.append(intervals[i])

            i += 1

        while i < n and intervals[i][0] <= new_interval[1]:
            new_interval[0] = min(new_interval[0], intervals[i][0])
            new_interval[1] = max(new_interval[1], intervals[i][1])

            i += 1

        merged.append(new_interval)

        while i < n:
            merged.append(intervals[i])

            i += 1

        return merged
